server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/UndesiredAllocationsTracker.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.common.FrequencyCappedAction;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.time.TimeProvider;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.core.TimeValue.ONE_MINUTE;
+import static org.elasticsearch.core.TimeValue.ZERO;
+import static org.elasticsearch.core.TimeValue.timeValueMillis;
+import static org.elasticsearch.core.TimeValue.timeValueMinutes;
+
+/**
+ * Keeps track of a limited number of shards that are currently in undesired allocations. If the
+ * shards remain in undesired allocations for longer than a configurable threshold, it will log
+ * why those shards can't be allocated to desired nodes.
+ */
+public class UndesiredAllocationsTracker {
+
+    private static final Logger logger = LogManager.getLogger(UndesiredAllocationsTracker.class);
+
+    private static final TimeValue FIVE_MINUTES = timeValueMinutes(5);
+
+    /**
+     * Warning logs will be periodically written if we see a shard that's been in an undesired allocation for this long
+     */
+    public static final Setting<TimeValue> UNDESIRED_ALLOCATION_DURATION_LOG_THRESHOLD_SETTING = Setting.timeSetting(
+        "cluster.routing.allocation.desired_balance.undesired_duration_logging.threshold",
+        FIVE_MINUTES,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
```

Comment:
Should this have a reasonable min?

Reply:
Added in 566f5d5699848d510eb98a24b2c0bb4183422a8e